README.md

```diff
@@ -511,6 +534,721 @@ will essentially perform a
+> Feedback on this question is welcome!
+
+#### A transactional lock-free leftist heap
+
+Let's implement something a bit more complicated,
+[a leftist heap](https://en.wikipedia.org/wiki/Leftist_tree), which is a kind of
+priority queue.
+
+> The implementation here is adapted from the book _Data Structures and
+> Algorithm Analysis in C (2nd ed.)_ by Mark Allen Weiss.
+
+First we define a data type to represent the spine of a leftist heap:
+
+```ocaml
+# type 'v leftist =
+    [ `Null
+    | `Node of 'v leftist Loc.t
+             * int Loc.t
+             * 'v
+             * 'v leftist Loc.t ]
+type 'v leftist =
+    [ `Node of 'v leftist Loc.t * int Loc.t * 'v * 'v leftist Loc.t | `Null ]
+```
+
+To create a leftist heap we
+[`make`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Loc/index.html#val-make)
+a location with an empty spine:
+
+```ocaml
+# let leftist () : _ leftist Loc.t = Loc.make `Null
+val leftist : unit -> 'a leftist Loc.t = <fun>
+```
+
+We then define an auxiliary function `npl_of` to get the null path length of a
+leftist heap:
+
+```ocaml
+# let npl_of ~xt : _ leftist -> int = function
+    | `Null -> 0
+    | `Node (_, npl, _, _) -> Xt.get ~xt npl
+val npl_of : xt:'a Xt.t -> 'b leftist -> int = <fun>
+```
+
+Notice the `~xt` parameter. It refers to the transaction log being passed
+explicitly. Above we pass it to
+[`get`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Xt/index.html#val-get)
+to record an operation in the log.
+
+The core operation of leftist heaps is that of merging two leftist heaps:
+
+```ocaml
+# let rec merge ~xt ~lt h1 h2 =
+    match h1, h2 with
+    | `Null, h2 -> h2
+    | h1, `Null -> h1
+    | (`Node (_, _, v1, _) as h1),
+      (`Node (_, _, v2, _) as h2) ->
+      let (`Node (h1l, npl, _, h1r) as h1), h2 =
+        if lt v1 v2 then h1, h2 else h2, h1 in
+      let l = Xt.get ~xt h1l in
+      if l == `Null then
+        Xt.set ~xt h1l h2
+      else begin
+        let r = merge ~xt ~lt (Xt.get ~xt h1r) h2 in
+        match npl_of ~xt l, npl_of ~xt r with
+        | l_npl, r_npl when l_npl < r_npl ->
+          Xt.set ~xt h1l r;
+          Xt.set ~xt h1r l;
+          Xt.set ~xt npl (l_npl + 1)
+        | _, r_npl ->
+          Xt.set ~xt h1r r;
+          Xt.set ~xt npl (r_npl + 1)
+      end;
+      h1
+val merge :
+  xt:'a Xt.t ->
+  lt:('b -> 'b -> bool) -> 'b leftist -> 'b leftist -> 'b leftist = <fun>
+```
+
+Again, `merge` passes the `~xt` parameter explicitly to the
+[`get`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Xt/index.html#val-get)
+and
+[`set`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Xt/index.html#val-set)
+operations to record them in the log.
+
+The `merge` operation can be used to implement both insertion to
+
+```ocaml
+# let insert ~xt ~lt h x =
+    let h1 = `Node (
+        Loc.make `Null,
+        Loc.make 1,
+        x,
+        Loc.make `Null
+      ) in
+    Xt.set ~xt h (merge ~xt ~lt h1 (Xt.get ~xt h))
```

Comment:
I  tried that : 
```ocaml 
let insert ~xt ~lt h x =
  let h1 = `Node (Loc.make `Null, Loc.make 1, x, Loc.make `Null) in
  Xt.modify ~xt h (fun hv -> merge ~xt ~lt h1 hv)
```
and it does not work as expected. The following test returns `[5]` instead of the whole list `[ 3; 1; 4; 1; 5 ]` : 
```ocaml
let test () =
  let lt = ( > ) in
  let a_heap = leftist () in
  Xt.commit
    { tx = (fun ~xt -> List.iter (insert ~xt ~lt a_heap) [ 3; 1; 4; 1; 5 ]) };
  Xt.commit { tx = (fun ~xt -> remove_all ~xt ~lt a_heap) }
```

Whereas this following equivalent works:   
```ocaml
let insert ~xt ~lt h x =
  let h1 = `Node (Loc.make `Null, Loc.make 1, x, Loc.make `Null) in
  let hv = Xt.get ~xt h in
  Xt.set ~xt h (merge ~xt ~lt h1 hv)
````

Note : `update` does not seem to work as well.

Reply:
Thanks!  I added [a note on this to PR#30](https://github.com/ocaml-multicore/kcas/pull/30#issuecomment-1461950511).